README.md
# mgtron_test_suite


CHANGELOG.md
# CHANGELOG

Author: Hunter, Christerpher

All notable changes will be appended here.

This project, henceforth, will recongnize [semantic versioning](https://semver.org/).

## [⭐.✴️.✳️] - YEAR MONTH DAY

Here we write upgrade and change notes.

⭐ MAJOR version when you make incompatible API changes,

✴️ MINOR version when you add functionality in a backwards compatible manner

✳️ PATCH version when you make backwards compatible bug fixes.

---

## ✴️[0.3.0] - 2023 JAN 10

- Added a quick test function

## ✴️[0.2.1] - 2023 JAN 06

- Init CHANGELOG
- corrected device path


rigol-plot.py
#!/usr/bin/env python

"""
Download data from a Rigol DS1052E oscilloscope and graph with matplotlib.
By Ken Shirriff, http://righto.com/rigol
Based on http://www.cibomahto.com/2010/04/controlling-a-rigol-oscilloscope-using-linux-and-python/
by Cibo Mahto.
"""

import numpy
import matplotlib.pyplot as plot
import sys
import pyvisa

rm = pyvisa.ResourceManager()
# Get the USB device, e.g. 'USB0::0x1AB1::0x0588::DS1ED141904883'
instruments = rm.list_resources()
usb = list(filter(lambda x: 'USB' in x, instruments))
if len(usb) != 1:
    print('Bad instrument list', instruments)
    sys.exit(-1)
# bigger timeout for long mem
scope = rm.open_resource(usb[0], timeout=20, chunk_size=1024000)

# Grab the raw data from channel 1
scope.write(":STOP")

# Get the timescale
timescale = float(scope.query(":TIM:SCAL?"))

# Get the timescale offset
timeoffset = float(scope.query(":TIM:OFFS?")[0])
voltscale = float(scope.query(':CHAN1:SCAL?')[0])

# And the voltage offset
voltoffset = float(scope.query(":CHAN1:OFFS?")[0])

scope.write(":WAV:POIN:MODE RAW")
rawdata = scope.query(":WAV:DATA? CHAN1").encode('ascii')[10:]
data_size = len(rawdata)
sample_rate = scope.query(':ACQ:SAMP?')[0]
print('Data size:', data_size, "Sample rate:", sample_rate)

scope.write(":KEY:FORCE")
scope.close()

data = numpy.frombuffer(rawdata, 'B')

# Walk through the data, and map it to actual voltages
# This mapping is from Cibo Mahto
# First invert the data
data = data * -1 + 255

# Now, we know from experimentation that the scope display range is actually
# 30-229.  So shift by 130 - the voltage offset in counts, then scale to
# get the actual voltage.
data = (data - 130.0 - voltoffset/voltscale*25) / 25 * voltscale

# Now, generate a time axis.
time = numpy.linspace(timeoffset - 6 * timescale,
                      timeoffset + 6 * timescale, num=len(data))

# See if we should use a different time axis
if (time[-1] < 1e-3):
    time = time * 1e6
    tUnit = "uS"
elif (time[-1] < 1):
    time = time * 1e3
    tUnit = "mS"
else:
    tUnit = "S"

# Plot the data
plot.plot(time, data)
plot.title("Oscilloscope Channel 1")
plot.ylabel("Voltage (V)")
plot.xlabel("Time (" + tUnit + ")")
plot.xlim(time[0], time[-1])
plot.show()


src/test_suite/rigol.py
from typing import Literal
from colorama import Fore as F
import pyvisa

R = F.RESET


class DSA800:
    """The commands that go directly to the Rigol DSA832E analyzer"""

    def __init__(self) -> None:
        self.name = "DS800Series"
        self.version = "0.2.0"
        self.manufacturer = "Rigol"
        self.rm = pyvisa.ResourceManager()  # Create a new resource manager object
        self.analyzer = self.rm.open_resource(
            resource_name='TCPIP0::10.10.30.50::INSTR',  # DSA832E
            timeout=5_000,
            chunk_size=1_024_000,
            access_mode=pyvisa.constants.AccessModes.no_lock,  # type: ignore
            read_termination='\n',
            write_termination='\n',
        )

    def _connect(self) -> list[str]:
        print(f"\n{F.YELLOW}Connecting to {self.name}...{R}\n")

        # read the USB device, e.g. 'USB0::0x1AB1::0x0588::DS1ED141904883'
        instruments = self.rm.list_resources()
        usb = list(filter(lambda x: 'USB' in x, instruments))
        if len(usb) != 1:
            print('Bad instrument list', instruments)
            exit(-1)

        return usb

    def _identification(self) -> str:
        """read the analyzer identification"""

        name = self.analyzer.query(":*IDN?")  # type: ignore

        name = name.split(",")[1]

        return name

    def establish_resource(self) -> pyvisa.resources.Resource:
        """Establish connection to the analyzer"""

        # bigger timeout for long mem
        self.analyzer = self.rm.open_resource(
            self._connect()[-1], timeout=5000, chunk_size=1024000)

        print(f"{F.GREEN}Connected to {self._identification()} {R}")

        return self.analyzer

    def read_start_frequency(self, freq: float | None = None) -> float:
        """Read or set start frequency in Hz"""

        # self.analyzer = self.establish_resource()

        frequency = self.analyzer.query(":FREQ:STAR?\r")  # type: ignore

        frequency = frequency if freq is None else self.analyzer.write(  # type: ignore
            f":FREQ:STAR {freq}")

        return frequency

    def read_stop_frequency(self, freq: float | None = None) -> float:
        """Read or set stop frequency in Hz"""
        frequency = self.analyzer.query(":FREQ:STOP?")  # type: ignore

        frequency = frequency if freq is None else self.analyzer.write(  # type: ignore
            f":FREQ:STOP {freq}")

        return frequency

    def write_center_frequency(self, freq: float) -> None:
        """Set center frequency in Hz"""

        self.analyzer.write(f":FREQ:CENT: {freq}")  # type: ignore

    def span(self, span: float) -> None:
        """Set span frequency in Hz"""

        self.analyzer.write(f":SENS:FREQ:SPAN {span}")  # type: ignore

    def write_bandwidth(self, bandwidth: float) -> None:
        self.analyzer.write(f":CALC:BAND:NBD {bandwidth}")  # type: ignore

    def read_bandwidth(self) -> str:
        return self.analyzer.write(":CALC:BAND:NBD?")  # type: ignore

    def enable_disable_tracking_generator(self, enable: bool) -> str:
        val = self.analyzer.write(  # type: ignore
            ":OUTP:STAT ON") if enable else self.analyzer.write(":OUTP:STAT  OFF")  # type: ignore

        return f"AFFIRMATIVE: {enable}" if val == 15 else "NEGATIVE"

    def set_power_output(self, power: int) -> float:
        """Oputput power: -40 to 0 dBm"""

        return float(
            self.analyzer.write(  # type: ignore
                f":SOUR:POW:LEV:IMM:AMPL {power}"
            )
        )

    def read_power_level(self) -> float | str:
        return (self.analyzer.query(":READ:CHP?"))  # type: ignore

    def disable_all_markers(self) -> str:
        """Disable all markers"""

        return self.analyzer.write(":CALC:MARK:AOFF")  # type: ignore

    def set_markers(self, marker: int, freq: float) -> None:
        """Set marker frequency"""

        match marker:

            case 1:
                self.analyzer.write(f":CALC:MARK1:MODE POS")  # type: ignore
                self.analyzer.write(f":CALC:MARK1:FUNC NDB")  # type: ignore
                self.analyzer.write(f":CALC:MARK1:STAT ON")  # type: ignore
                self.analyzer.write(f":CALC:MARK1:X {freq}")  # type: ignore

            case 2:
                self.analyzer.write(f":CALC:MARK2:FUNC NDB")  # type: ignore
                self.analyzer.write(f":CALC:MARK2:MODE POS")  # type: ignore
                self.analyzer.write(f":CALC:MARK2:STAT ON")  # type: ignore
                self.analyzer.write(f":CALC:MARK2:X {freq}")  # type: ignore

            case 3:
                self.analyzer.write(f":CALC:MARK3:FUNC NDB")  # type: ignore
                self.analyzer.write(f":CALC:MARK3:MODE POS")  # type: ignore
                self.analyzer.write(f":CALC:MARK3:STAT ON")  # type: ignore
                self.analyzer.write(f":CALC:MARK3:X {freq}")  # type: ignore

    def continuous_peak(self, on_or_off: Literal["ON"] | Literal["OFF"]) -> None:
        """Turn on or off continuous peak"""

        self.analyzer.write(  # type: ignore
            f":CALC:MARK1:CPE:STAT {on_or_off.capitalize()}")

    def peak_span_center(self, on_or_off: Literal["ON"] | Literal["OFF"]) -> None:
        """Turn on or off peak span center"""

        self.analyzer.write(":CALC:MARK1:PEAK:SET:CF")  # type: ignore

    def save_trace(self, trace_num: int, filename: str) -> None:
        """Fetch the trace data from the instrument"""

        self.analyzer.write(  # type: ignore
            f":MMEM:STOR:TRAC TRACE{trace_num}, E:Trace{trace_num}:{filename}.csv")

    def get_trace_data(self, trace_num: int) -> str:
        """Fetch the trace data from the instrument"""

        return self.analyzer.query(  # type: ignore
            f":TRAC:DATA? TRACE{trace_num}"
        )

    def screen_state(self, on_or_off: int) -> None:
        """Turn the screen on or off; Off increases measurement speed"""

        self.analyzer.write(f":DISP:ENAB {on_or_off}")  # type: ignore

    def save_screenshot(self, filename: str) -> None:
        """Save screenshot to file on usb drive"""

        self.analyzer.write(  # type: ignore
            f":MMEM:STOR:SCR E:{filename}.png")

    def save_csv(self, filename: str) -> None:
        """Save CSV to file on usb drive"""

        self.analyzer.write(  # type: ignore
            f":MMEM:STOR:RES E:{filename}.csv")

    def read_instrument(self, session) -> str:
        """Read instrument"""

        return self.analyzer.visalib.read(  # type: ignore
            session=session,
            count=1000
        )

    def half_span(self) -> None:
        """Set half span"""

        self.analyzer.write(":SENS:FREQ:SPAN:ZIN")  # type: ignore

    def full_span(self) -> None:
        """Set full span"""

        self.analyzer.write(":SENS:FREQ:SPAN:ZOUT")  # type: ignore


src/test_suite/interface.py
from dataclasses import dataclass
import pathlib
import platform
from time import sleep
import serial
import subprocess
import logging


ROOT = pathlib.Path(__file__).resolve().parent.parent.parent

logger = logging.getLogger(__name__)
logging.basicConfig(
    level=logging.DEBUG,
    format="%(asctime)s :: %(name)s :: %(message)s :: %(levelname)s",
    datefmt="%d-%b-%y %H:%M:%S",
    filename=f"{ROOT}/mg.log",
    filemode="w",
)


BAUDRATE = 115_200
DEVICE_PORT: int = int()
# PORT: str = str()


def find_device(DEVICE_NUMBER: int = DEVICE_PORT) -> tuple[str, list[str]]:
    """Find the Megatron device plugged into the Linux computer"""

    # Determine if system is Linux or WIN
    if platform.system().lower() == "linux":

        # Search Linux filesystem for device
        find = ["find /dev -iname '*acm*'"]
        try:
            logger.info(
                msg=f"{find_device.__name__} function executing linux shell command to get device file names"
            )
            results_ = subprocess.run(
                args=find,
                shell=True,
                stdout=subprocess.PIPE,
                universal_newlines=True,
                encoding="utf-8",
                capture_output=False,
            )
            results = sorted(results_.stdout.strip().splitlines())
            # global PORT
            PORT = results[DEVICE_NUMBER]
            logger.info(msg=f"Connected Devices: {results}")
            logger.debug(msg=f"Chosen Device: {PORT}")
            return results[DEVICE_NUMBER], results
        except IndexError:
            logger.exception("Device not found")

    elif platform.system().lower() == "windows":
        logger.info(f"{platform.system()} discovered")
        # global PORT
        # Search Windows filesystem for device
        # filename = "COM3"
        # devices = [os.path.join(root, filename) for root, dir,
        #           files in os.walk("/user/") if filename in files]

        return "COM3", ["COM3"]
    return "None", ["None"]


def serial_call(*args, PORT: str = find_device(DEVICE_NUMBER=0)[0][0]) -> None:
    logger.debug(msg="Serial call to device initiated")

    NAME: str = serial_call.__name__

    sleep(0.1)  # Allow time to read and execute via serial *REQUIRED*

    try:
        with serial.Serial() as ser:  # If the teensy is corrupted the GUI freezes here
            ser.port = PORT.strip()
            logger.debug(
                msg=f"Serial device set from {NAME}()")
            ser.baudrate = BAUDRATE
            logger.debug(msg=f"baudrate set from {NAME}()")
            ser.timeout = 2  # seconds
            logger.debug(msg=f"timeout set from {NAME}()")
            ser.open()
            logger.debug(
                msg=f"opening serial connection from {NAME}()")
            ser.write(f"{' '.join([arg for arg in args])}".encode("utf-8"))
            logger.debug(
                msg=f"information sent via serial protocol; {NAME}()")
            ser.flush()

    except (serial.SerialException, NameError):
        logger.exception(msg="No device found")

    logger.debug(
        msg=f"Contextually closed serial connection from {NAME}()")


@dataclass(slots=True)
class Megatron:
    """Class to organize the manipulation of 8 channels"""

    logger.info(msg="\n\nGUI LAUNCHED\n\n")

    try:
        logger.info(msg=f"Getting the port name of device")
        PORT = find_device(DEVICE_PORT)[0]  # type: ignore
        logger.info(msg=f"Connected device path: {PORT}")
    except TypeError:
        logger.exception(msg="No device found on system")

    @classmethod
    def status(cls, PORT: str) -> None:
        """Check the status of the board"""

        logger.info(f"{Megatron.status.__name__} function executed")
        serial_call("s", PORT=PORT)

    def change_power(self, channel: int, power_level: int, PORT: str):
        """
        Change the power level of a channel
        Range: 0 - 63
        """
        print(f"Change power: PORT === {PORT}")
        logger.info(f"{Megatron.change_power.__name__} function executed")
        logger.info(msg=f"Connected device path: {PORT}")
        return serial_call("p", str(channel), str(power_level), PORT=PORT)

    def change_freq(self, channel: int, frequency: float, PORT: str):
        """
        Change the frequency of a channel
        Range: 50 - 6400 MHz
        """

        logger.info(f"{Megatron.change_freq.__name__} function executed")
        # print()
        return serial_call("f", str(channel), str(frequency), PORT=PORT)

    def change_bandwidth(self, channel: int, percentage: int, PORT: str):
        """
        Change the bandwidth of a channel
        Range: 0 - 100
        """

        logger.info(f"{Megatron.change_bandwidth.__name__} function executed")
        return serial_call("b", str(channel), str(percentage), PORT=PORT)

    def save_state(self, state: bool, PORT: str) -> None:
        """Save each settings made by the user into memory for next startup"""

        state = 1 if state else 0  # type: ignore
        try:
            serial_call("x", str(state), PORT=PORT)
            logger.info(f"{Megatron.save_state.__name__} function executed")
        except TypeError:
            logger.exception(msg="No device assigned")

    def amplification(self, channel: int, state: bool, PORT: str) -> None:
        """Output HIGH or LOW logic level out of a chosen channel"""

        state = 1 if state else 0  # type: ignore
        serial_call("a", str(channel), str(state), PORT=PORT)
        logger.info(f"{Megatron.amplification.__name__} function executed")

    def stability(self, state: bool, PORT: str) -> None:
        """Boolean a second filtering stage of capacitors for further stability"""

        state = 1 if state else 0  # type: ignore
        serial_call("~", str(state), PORT=PORT)
        logger.info(f"{Megatron.stability.__name__} function executed")

    def noise_control(self, state: bool, percentage: int, PORT: str) -> None:
        """
        Optimal settings hardcoded; Input @ %100 Output @ %85
        state 0: Output stage
        state 1: Input stage
        """

        state = 1 if state else 0  # type: ignore
        serial_call("n", str(state), str(percentage), PORT=PORT)
        logger.info(f"{Megatron.noise_control.__name__} function executed")

    def reset_board(self, PORT: str) -> None:
        """Reset the parameters of the board"""

        [
            (
                serial_call("p", str(i), "0", PORT=PORT),
                # serial_call("b", str(i), "0", PORT=PORT),
                # serial_call("f", str(i), "50.00", PORT=PORT),
            )
            for i in range(1, 9)
        ]

        logger.info(f"{Megatron.reset_board.__name__} function executed")

    logger.info(msg="class Megatron initialized")


logger.debug(msg="EOF")


def main() -> None:
    import random

    # find_device("linux")
    # test_1 = Megatron()

    # for i in range(8):
    # test_1.change_power(i+1, random.randint(a=10, b=63))
    # sleep(1)
    # test_1.change_freq(i+1, random.randint(a=50, b=6300))
    # sleep(1)
    # test_1.change_bandwidth(i+1, random.randint(a=10, b=100))
    # sleep(1)
    # sleep(1)
    # test_1.reset_board()

    # test_1.change_freq(1, 2545.54)
    # test_1.change_power(1, 63)

    # test_1.status()

    # test_1.amplification(3, True)
    # test_1.stability(True)
    # test_1.save_state(True)


if __name__ == "__main__":
    main()


src/test_suite/mgtron_tester.py
from src.test_suite.interface import Megatron
import time
import random
from colorama import Fore as F

R = F.RESET


def random_frequency_test():
    """Randomly loops through every value from 50 Mhz to 6Ghz"""
    frequency_list = []
    for i in range(50, 6001, 500):
        frequency_list.append(i)
    # random.shuffle(frequency_list)
    return frequency_list


def random_power_test():
    """Randomly loops through every value from 0 to 63"""
    power_list = []
    for i in range(0, 64, 2):
        power_list.append(i)
    # random.shuffle(power_list)
    return power_list


def random_bandwidth_test():
    """Randomly loops through every value from 0 to 100"""
    bandwidth_list = []
    for i in range(0, 101, 10):
        bandwidth_list.append(i)
    random.shuffle(bandwidth_list)
    return bandwidth_list


def channel_select() -> list[int]:
    """Select channel one through eight"""
    channel: list = [x for x in range(1, 9)]
    return channel


def random_test_suite():
    """create every possible combination of frequency, power, and bandwidth"""

    mgtron = Megatron()
    frequency_list = random_frequency_test()
    power_list = random_power_test()
    # bandwidth_list = random_bandwidth_test()
    channel_list = channel_select()
    for channel in channel_list:
        for frequency in frequency_list:
            for power in power_list:

                mgtron.change_freq(
                    frequency=frequency, channel=channel)
                mgtron.change_power(
                    power_level=power, channel=channel)
                mgtron.change_bandwidth(
                    percentage=0, channel=channel)
                time.sleep(0.1)
                print(
                    f"Channel: {F.YELLOW}{channel}{R}, Frequency: {F.YELLOW}{frequency}{R}, Power: {F.CYAN}{power}{R}")


def main():

    print(f"{F.GREEN}Welcome to the MGTron Tester{R}")

    begin_time = time.time()

    mgtron = Megatron()
    frequency_list = random_frequency_test()
    power_list = random_power_test()
    # bandwidth_list = random_bandwidth_test()
    channel_list = channel_select()
    for channel in channel_list:
        for frequency in frequency_list:
            for power in power_list:

                mgtron.change_freq(
                    frequency=frequency, channel=channel)
                mgtron.change_power(
                    power_level=power, channel=channel)
                mgtron.change_bandwidth(
                    percentage=0, channel=channel)
                time.sleep(0.1)
                print(
                    f"Channel: {F.YELLOW}{channel}{R}, Frequency: {F.YELLOW}{frequency}{R}, Power: {F.CYAN}{power}{R}")

    end_time = time.time()
    print(f"\n{F.RED}Time Elapsed: {(end_time - begin_time) / 60:.2f}{R}")


if __name__ == "__main__":
    main()


main.py
import glob
from typing import Iterable
from src.test_suite.interface import Megatron
from src.test_suite.rigol import DSA800
import time
import random
from colorama import Fore as F
import pandas as pd
import matplotlib.pyplot as plt

R = F.RESET


def random_frequency_test():
    """Randomly loops through every value from 50 Mhz to 6Ghz"""
    frequency_list = []
    for i in range(50, 6001, 500):
        frequency_list.append(i)
    # random.shuffle(frequency_list)
    return frequency_list


def random_power_test():
    """Randomly loops through every value from 0 to 63"""
    power_list = []
    for i in range(0, 64, 2):
        power_list.append(i)
    # random.shuffle(power_list)
    return power_list


def random_bandwidth_test():
    """Randomly loops through every value from 0 to 100"""
    bandwidth_list = []
    for i in range(0, 101, 10):
        bandwidth_list.append(i)
    random.shuffle(bandwidth_list)
    return bandwidth_list


def channel_select() -> list[int]:
    """Select channel one through eight"""
    channel: list = [x for x in range(1, 9)]
    return channel


def random_test_suite():
    """create every possible combination of frequency, power, and bandwidth"""

    mgtron = Megatron()
    frequency_list = random_frequency_test()
    power_list = random_power_test()
    # bandwidth_list = random_bandwidth_test()
    channel_list = channel_select()
    for channel in channel_list:
        for frequency in frequency_list:
            for power in power_list:

                mgtron.change_freq(
                    frequency=frequency, channel=channel, PORT="/dev/ttyACM0")
                mgtron.change_power(
                    power_level=power, channel=channel, PORT="/dev/ttyACM0")
                mgtron.change_bandwidth(
                    percentage=0, channel=channel, PORT="/dev/ttyACM0")
                time.sleep(0.1)
                print(
                    f"Channel: {F.YELLOW}{channel}{R}, Frequency: {F.YELLOW}{frequency}{R}, Power: {F.CYAN}{power}{R}")


def automate_octo_test(band: int) -> dict[str, float]:

    # evenly space dict values from 2400 to 2800 over 8 points
    two_four: dict[str, float] = {
        "1": 2400,
        "2": 2450,
        "3": 2500,
        "4": 2550,
        "5": 2600,
        "6": 2650,
        "7": 2700,
        "8": 2750,
    }

    # Put eight 5 Ghz band channels into a list
    five_eight: dict[str, float] = {
        "1": 5000,
        "2": 5150,
        "3": 5200,
        "4": 5250,
        "5": 5300,
        "6": 5350,
        "7": 5400,
        "8": 5450,
    }

    if band == 2:
        return two_four

    elif band == 5:
        return five_eight

    else:
        print("Invalid Band")
        return {}


def kill_power(mgtron: Megatron, port: int):
    """Kill the power to all of the channels"""

    for i in range(1, 9):

        mgtron.change_freq(
            frequency=50,
            channel=i,
            PORT=f"/dev/ttyACM{port}",
        )

        mgtron.change_power(
            power_level=0,
            channel=i,
            PORT=f"/dev/ttyACM{port}",
        )

        mgtron.change_bandwidth(
            percentage=0,
            channel=i,
            PORT=f"/dev/ttyACM{port}",
        )


def quick_test(mgtron: Megatron, port: int, channel: int, rigol: DSA800) -> None:
    """Quick test of the megatron"""

    window_offet: float = 500e6
    conv_factor: float = 1e6

    quick_freqs: list = [
        1000,
        2000,
        3000,
        4000,
        5000,
        6000,
    ]

    for freq in quick_freqs:

        # rigol.write_center_frequency(freq=freq * conv_factor)
        # time.sleep(1)

        # rigol.span(span=window_offet)

        # Set the peak measurement to channel variable
        print(f"\n{F.GREEN}Setting Peak Measurement{R}")
        rigol.set_markers(marker=1, freq=freq * conv_factor)
        time.sleep(0.1)

        mgtron.change_freq(
            frequency=freq,
            channel=channel,
            PORT=f"/dev/ttyACM{port}",
        )

        mgtron.change_power(
            power_level=63,
            channel=channel,
            PORT=f"/dev/ttyACM{port}",
        )

        # Should already be zero
        mgtron.change_bandwidth(
            percentage=0,
            channel=channel,
            PORT=f"/ dev/ttyACM{port}"
        )

        # Take rigol trace values
        # print(f"{F.GREEN}Taking Rigol Screenshot{R}")
        # rigol.save_trace(
        #     trace_num=1, filename=f"port_{port}_channel_{channel}_freq_{freq}")

        # time.sleep(3)

        # print(rigol.read_instrument(DSA800._connect))

        time.sleep(0.1)


def wifi_test(mgtron: Megatron, port: int, channel: int, rigol: DSA800, wifi_freqs: list[float]) -> None:
    """Test the wifi frequencies"""

    window_offet: float = 500e6
    conv_factor: float = 1e6

    for freq in wifi_freqs:

        # rigol.write_center_frequency(freq=freq * conv_factor)
        # time.sleep(1)

        # rigol.span(span=window_offet)

        # Set the peak measurement to channel variable
        print(f"\n{F.GREEN}Setting Peak Measurement{R}")
        rigol.set_markers(marker=1, freq=freq * conv_factor)
        time.sleep(0.1)

        mgtron.change_freq(
            frequency=freq,
            channel=channel,
            PORT=f"/dev/ttyACM{port}",
        )

        mgtron.change_power(
            power_level=63,
            channel=channel,
            PORT=f"/dev/ttyACM{port}",
        )

        # Should already be zero
        mgtron.change_bandwidth(
            percentage=0,
            channel=channel,
            PORT=f"/dev/ttyACM{port}"
        )

        # Take rigol trace values
        print(f"{F.GREEN}Taking Rigol Screenshot{R}")
        # rigol.save_trace(
        # trace_num=1, filename=f"port_{port}_channel_{channel}_freq_{freq}")

        print(rigol.get_trace_data(trace_num=1))

        # rigol.save_screenshot(
        # filename=f"port_{port}_channel_{channel}_freq_{freq}")

        time.sleep(4)


def plot_csv(filename: str) -> None:
    """Plot the csv file"""

    df = pd.read_csv(filename, skiprows=3)

    df.columns.values[:3] = ["Frequency", "", "Amplitude"]

    # Legend
    plt.legend(["Amplitude"])

    # Mark the highest amplitutde peak in the trace with a long arrow
    plt.annotate(
        f"Peak Amplitude: {df['Amplitude'].max()} dB\nFrequency: {df['Frequency'][df['Amplitude'].idxmax()] // 1e6:.2f} MHz",
        xy=(df["Frequency"][df["Amplitude"].idxmax()],
            df["Amplitude"].max()),
        xytext=(df["Frequency"][df["Amplitude"].idxmax()],
                df["Amplitude"].max() - 5),
        arrowprops=dict(facecolor="red", shrink=0.05),
    )

    # extract channel and port from '7/0286-5B07/TRACE1:PORT_0_CHANNEL_1_FREQ_1000.CSV.png'
    port = filename.split(":")[1].split("_")[:2][-1]
    channel = filename.split(":")[1].split("_")[3]
    freq = filename.split(":")[1].split("_")[5].split(".")[0]

    print(f"{F.GREEN}Saving plot{R}")
    print(f"{F.GREEN}Port: {port}, Channel: {channel}, Frequency: {freq}{R}")

    plt.xlabel("Frequency (MHz)")
    plt.ylabel("Amplitude (dBm)")
    plt.title(f"Card: {port}, Channel: {channel}, Frequency: {freq}")
    plt.plot(df["Frequency"], df["Amplitude"])
    plt.savefig(f"trace_plots/ACM{port}_channel_{channel}_freq_{freq}.png")
    plt.close()


def main():

    print(f"{F.GREEN}Welcome to the MGTron Tester{R}")

    mgtron = Megatron()
    rigol = DSA800()

    port = int(input("Enter enuermated device number: "))
    band_to_test = int(input("Enter band to test, 2 for 5: "))

    begin_time = time.time()

    band = automate_octo_test(band_to_test)

    # Get all of the files in /run/media/djhunter67/0286-5B07/ that end with .CSV
    # files = glob.glob("/run/media/djhunter67/0286-5B07/*.CSV")

    # [
    #     plot_csv(file) for file in files
    # ]

    [
        (
            kill_power(port=port, mgtron=mgtron),
            wifi_test(
                mgtron=mgtron,
                port=port,
                channel=channel,
                rigol=rigol,
                wifi_freqs=[
                    freq for freq in band.values()
                ]
            ),
        ) for channel in range(1, 9)

    ]
    end_time = time.time()
    print(f"\n{F.RED}Time Elapsed: {(end_time - begin_time) / 60:.2f}{R} minutes")
    kill_power(port=port, mgtron=mgtron)


if __name__ == "__main__":
    main()
